Filter PRs based on repo label

RascalTwo opened the pull request:
Filters out PRs based on the [new rules](https://hacktoberfest.digitalocean.com/hacktoberfest-update), implemented by digitalocean/hacktoberfest#596.

These new rules are only applied to PRs created after `2020-10-03T12:00:00.000Z`.

***

For all PRs created after the above date, their repos labels are fetched and a `repo_has_hacktoberfest_topic` property is added to the PR, representing if the PR repo has the `hacktoberfest` topic.

Then PRs that are created after the above date and their repo doesn't have the `hacktoberfest` topic are filtered out.

***

The pending time has been increased from 7 days to 14 days.

The `has_hacktoberfest_label` now represents if the `hacktoberfest-accepted` label is present, instead of the `hacktoberfest` label.

***

I haven't used Lodash in a while, so hopefully I've implemented this in the proper style, I'm more familiar with the native variants of its methods.

jenkoian commented:
Testing results:

* PRs prior to October 3, 2020 at 12:00:00 UTC ✅ 
* Submitted in a repo with the hacktoberfest topic AND
during the month of October AND The PR is merged ❌ 
* Submitted in a repo with the hacktoberfest topic AND
during the month of October AND The PR is labelled as hacktoberfest-accepted by a maintainer ✅ 
* Submitted in a repo with the hacktoberfest topic AND
during the month of October AND The PR has been approved ✅ 

Are you able to look into the merged thing? Thanks so much for this btw.

RascalTwo commented:
Hey, thanks for taking the time to review it, glad it's mostly right!

***

So I did some more extensive testing, I was able to find a bug - I don't think it's related to what you found though, it's due to the fact that the `created_at` provided by the GitHub API are UTC, and they're being parsed with `moment()`, not `moment.utc()`.

I did random manual testing, also some [quick and dirty jest tests](https://github.com/RascalTwo/hacktoberfest-checker/blob/filter-prs-based-on-repo-label-testing/api/controllers/pr/findPrs/index.test.js), neither of which allowed me to replicate the

- Submitted in a repo with the hacktoberfest topic AND
during the month of October AND The PR is merged :x: 

problem.

> If you want, you can clone it and run `npx jest` to see the results yourself

***

Perhaps you could share what usernames you looked up/steps you took to produce the problem?

jenkoian commented:
@RascalTwo so if I check out your branch and search for your username I wouldn't expect this PR to show as although it's in a repo with the #hacktoberfest label, it's not merged, approved or has the #hacktoberfest-accepted label. That's how I tested that. Does that make sense?

RascalTwo commented:
Ah I see what you mean, got mixed up as originally the `merged` status wasn't used for filtering, but now both `merged` and `approved` do filter under the new rules.

Updated accordingly, thanks for taking the time to clarify it for me!

jenkoian commented on api/controllers/pr/findPrs/index.js:
```diff
@@ -76,7 +124,37 @@ const findPrs = (github, username) => {
+          pr.merged ||
+          pr.approved ||
+          pr.has_hacktoberfest_label
+      )
```
Not tested this but just looking, shouldn't it be...

```javascript
pr.has_hacktoberfest_label || 
(	
	pr.repo_must_have_topic && 
	(pr.merged || pr.approved)
)
```

> The PR is labelled as hacktoberfest-accepted by a maintainer OR
  Submitted in a repo with the hacktoberfest topic AND (
    The PR is merged OR
    The PR has been approved
  )

RascalTwo replied:
Oh didn't realize there was a 3rd update to the rules, I've updated the code to match these latest rules.

RascalTwo commented:
Updated the filters to match the 3rd rule change, also made the filters more readable.

jenkoian reviewed (approved):
Tested all scenarios and it seems to work really nicely.

jenkoian commented:
Approved so this is optional, but don't suppose you want to update the FAQ to reflect this change? https://github.com/jenkoian/hacktoberfest-checker/blob/master/src/pages/Faq/data.js

RascalTwo commented:
Glad it's works, thanks for taking the time to review!

Updated the FAQ to not be in direct conflict with the latest rules; added `approved`, changed the time to two weeks, etc.
